src/Common/tests/System/Net/Http/LoopbackServer.cs

```diff
@@ -4,274 +4,213 @@
+                    await funcAsync(server);
                 }
-
-                var url = new Uri($"{scheme}://{host}:{localEndPoint.Port}/");
-
-                return funcAsync(server, url).ContinueWith(t =>
-                {
-                    server.Dispose();
-                    t.GetAwaiter().GetResult();
-                }, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default);
-            }
-            catch (Exception e)
-            {
-                localEndPoint = null;
-                return Task.FromException(e);
             }
         }
 
-        public static Task CreateServerAndClientAsync(Func<Uri, Task> clientFunc, Func<Socket, Task> serverFunc)
+        public static Task CreateServerAsync(Func<LoopbackServer, Uri, Task> funcAsync, Options options = null)
         {
-            return CreateServerAsync(async (server, uri) =>
+            return CreateServerAsync(server => funcAsync(server, server.Uri), options);
+        }
+
+        public static Task CreateClientAndServerAsync(Func<Uri, Task> clientFunc, Func<LoopbackServer, Task> serverFunc)
+        {
+            return CreateServerAsync(async server =>
             {
-                Task clientTask = clientFunc(uri);
+                Task clientTask = clientFunc(server.Uri);
                 Task serverTask = serverFunc(server);
 
                 await new Task[] { clientTask, serverTask }.WhenAllOrAnyFailed();
             });
         }
 
-        public static string DefaultHttpResponse => $"HTTP/1.1 200 OK\r\nDate: {DateTimeOffset.UtcNow:R}\r\nContent-Length: 0\r\n\r\n";
+        private static string GetDefaultHttpResponse() => $"HTTP/1.1 200 OK\r\nDate: {DateTimeOffset.UtcNow:R}\r\nContent-Length: 0\r\n\r\n";
 
-        public static IPAddress GetIPv6LinkLocalAddress() =>
-            NetworkInterface
-                .GetAllNetworkInterfaces()
-                .SelectMany(i => i.GetIPProperties().UnicastAddresses)
-                .Select(a => a.Address)
-                .Where(a => a.IsIPv6LinkLocal)
-                .FirstOrDefault();
-
-        public static Task<List<string>> ReadRequestAndSendResponseAsync(Socket server, string response = null, Options options = null)
+        public async Task<List<string>> AcceptConnectionSendResponseAndCloseAsync(string response)
         {
-            return AcceptSocketAsync(server, (s, stream, reader, writer) => ReadWriteAcceptedAsync(s, reader, writer, response), options);
-        }
+            List<string> lines = null;
 
-        public static async Task<List<string>> ReadWriteAcceptedAsync(Socket s, StreamReader reader, StreamWriter writer, string response = null)
-        {
-            // Read request line and headers. Skip any request body.
-            var lines = new List<string>();
-            string line;
-            while (!string.IsNullOrEmpty(line = await reader.ReadLineAsync().ConfigureAwait(false)))
+            // Note, we assume there's no request body.  
```

Comment:
Does that mean we can't use the LoopbackServer for testing POST requests?

Reply:
No, it means we can't use this particular method for POST requests.  Instead you'd use AcceptConnectionAsync + Connection.ReadRequestHeaderAsync or something along those lines.